[RFC] sample: shell: Add sample showing how to test shell

PerMac opened the pull request:
The tests are ment to be used with the existing shell module sample from Zephyr.

PerMac committed:
sample: shell: Add sample showing how to test shell

The tests are ment to be used with the existing shell module sample
from Zephyr.

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

gchwier commented on samples/shell/README.rst:
```diff
@@ -0,0 +1,144 @@
+
+`A checkouted Zephyr repository is needed to execute this sample.`
+
+#. Copy the source file of the test (`shell_test.py`) into a directory with the shell module sample (`zephyr/samples/subsys/shell/shell_module`).
```
you can also create `testspec.yaml` in `shell_test.py` directory, and point a `sourc_dir` (absolute path or erlated to ZEPHYR_BASE), e.g.
```
common:
  source_dir: samples/subsys/shell/shell_module
tests:
  sample.shell.shell_module:
     ...
```

PerMac replied:
Added

PerMac committed:
specification: Unify path resolution

Use zephyr_base in RegularSpecificationProcessor when
determining path to source directory

Signed-off-by: Maciej Perkowski <Maciej.Perkowski@nordicsemi.no>

PerMac commented on samples/shell/shell_test.py:
```diff
@@ -0,0 +1,41 @@
+    timeout = 15  # [sec]
+    time_end = time.time() + timeout
+    while time.time() < time_end:
+        line = connection.readline().decode('UTF-8').strip()
```
This was my first approach but it seems there is some issue with `iter_stdout` since it doesn't work reliably with qemu. I agree using it could be better since `handler.log` is connected with `iter_stdout`.

PerMac commented on samples/shell/shell_test.py:
```diff
@@ -0,0 +1,41 @@
+@pytest.mark.build_specification('sample.shell.shell_module')
+def test_shell_help_and_ping(builder, dut):
+    time.sleep(1)  # wait for application initialization on DUT
+    dut.connection.write(b'help\n')
```
Indeed. Since we don't have it yet it was mentioned in the README.rst, that `Bidirectional communication is not yet implemented for *native_posix*.`

hakehuang commented on samples/shell/shell_test.py:
```diff
@@ -0,0 +1,41 @@
+logger = logging.getLogger(__name__)
+
+
+def is_regex_found(regex, connection):
```
this looks like a common fucntion, maybe we need put to a common place like create a test_helpers module?

PerMac replied:
Yes. For now this is a first sample showing the idea. We believe providing a module with common helper functions that can be reused between tests will definitely be useful.

hakehuang commented on src/twister2/specification_processor.py:
```diff
@@ -37,6 +37,7 @@ class SpecificationProcessor(abc.ABC):
 
     def __init__(self, twister_config: TwisterConfig) -> None:
         self.twister_config = twister_config
+        self.zephyr_base: str = self.twister_config.zephyr_base
```
there is a problem at my side, whenever I want to use pytest, it always ask me to set the zephyr_base. even it is not a twister scripte, this occures to me like the twister plugin contaminate the pytest envrionment?

PerMac replied:
This shouldn't be the case. We modify the plugin in a way that it has to be explicitly turned on, by adding `--twister` to pytest call. To mitigate exactly such issues as you describes. It should only ask you for zephyr base if you use `--twister` in pytest call. Maybe you have some older version of the plugin installed?

hakehuang replied:
there is an older version, I remove them then it is fine now.

hakehuang commented on samples/shell/shell_test.py:
```diff
@@ -0,0 +1,41 @@
+import time
+import re
+import pytest
```
would this be more approriate to `import twister`

PerMac replied:
You have to import pytest. This way you have access to pytest module. After installing our plugin it became a part of pytest (when turned on with --twister). Hence you need pytest
